File: kirby.lua
local Kirby = {}
Kirby.__index = Kirby

function Kirby.new(game)
	local self = setmetatable({}, Kirby)

	self.game = game
	self.x = 100
	self.y = 180
	self.gravity = 0.8
	self.velocityY = 0
	self.ducking = false
	self.jumping = false
	self.offsetY = 0
	self.image = display.newImage(self.game.group, 'kirby.png')
	self.image.x = self.x
	self.image.y = self.y

	return self
end

function Kirby:duck()
	if (self.jumping == false) then
		self.ducking = true
	end
end

function Kirby:jump()
	if (self.ducking == false and self.jumping == false) then
		self.jumping = true
		self.velocityY = -1.15
	end
end

function Kirby:update(delta)
	if (self.ducking) then
		local offset = delta * 0.2

		if (self.game.input.left) then
			self.offsetY = math.min(64, self.offsetY + offset)
		else
			self.offsetY = math.max(0, self.offsetY - offset)
		end

		if (self.offsetY == 0) then
			self.ducking = false
			if (self.game.input.right) then
				self:jump()
			end
		end

	elseif (self.jumping) then
		self.offsetY = math.min(0, self.offsetY + self.velocityY * delta)
		self.velocityY = self.velocityY + self.gravity * delta * 0.005

		if (self.offsetY == 0) then
			self.jumping = false
			self.ducking = self.game.input.left
		end
	end

	self.image.y = self.y + self.offsetY
end

return Kirby

File: src/kirby.js
define(['assets', 'rect'], function (Assets, Rect) {
	function Kirby(game) {
		PIXI.Sprite.call(this, Assets.kirby);

		this.game = game;
		this.position.x = 100;
		this.startY = 148;
		this.position.y = this.startY;
		this.gravity = 0.8;
		this.velocityY = 0;
		this.ducking = false;
		this.jumping = false;
		this.offsetY = 0;
	}

	Kirby.prototype = Object.create(PIXI.Sprite.prototype);
	Kirby.prototype.constructor = Kirby;

	Kirby.prototype.duck = function () {
		if (!this.jumping) {
			this.ducking = true;
		};
	};

	Kirby.prototype.jump = function () {
		if (!this.ducking && !this.jumping) {
			this.jumping = true;
			this.velocityY = -1.15;
		};
	};

	Kirby.prototype.update = function (delta) {
		var offset = delta * 0.2;

		if (this.ducking) {
			if (this.game.input.left) {
				this.offsetY = Math.min(64, this.offsetY + offset);
			} else {
				this.offsetY = Math.max(0, this.offsetY - offset);
			}

			if (this.offsetY === 0) {
				this.ducking = false;
				if (this.game.input.right) {
					this.jump();
				}
			}

		} else if (this.jumping) {
			this.offsetY = Math.min(0, this.offsetY + this.velocityY * delta);
			this.velocityY = this.velocityY + this.gravity * delta * 0.005;

			if (this.offsetY === 0) {
				this.jumping = false;
				this.ducking = this.game.input.left;
			}
		}

		this.position.y = Math.floor(this.startY + this.offsetY);
	};

	return Kirby;
});
